keep/providers/providers_service.py

```diff
@@ -227,34 +227,49 @@ def update_provider(
     def delete_provider(
         tenant_id: str, provider_id: str, session: Session, allow_provisioned=False
     ):
-        provider = session.exec(
+        provider_model: Provider = session.exec(
             select(Provider).where(
                 (Provider.tenant_id == tenant_id) & (Provider.id == provider_id)
             )
         ).one_or_none()
 
-        if not provider:
+        if not provider_model:
             raise HTTPException(404, detail="Provider not found")
 
-        if provider.provisioned and not allow_provisioned:
+        if provider_model.provisioned and not allow_provisioned:
             raise HTTPException(403, detail="Cannot delete a provisioned provider")
 
         context_manager = ContextManager(tenant_id=tenant_id)
         secret_manager = SecretManagerFactory.get_secret_manager(context_manager)
+        config = secret_manager.read_secret(provider_model.configuration_key, is_json=True)
+        provider = ProvidersFactory.get_provider(
+                context_manager, provider_model.id, provider_model.type, config
+            )
 
         try:
-            secret_manager.delete_secret(provider.configuration_key)
+            secret_manager.delete_secret(provider_model.configuration_key)
         except Exception:
             logger.exception("Failed to delete the provider secret")
 
-        if provider.consumer:
+        if provider_model.consumer:
             try:
                 event_subscriber = EventSubscriber.get_instance()
-                event_subscriber.remove_consumer(provider)
+                event_subscriber.remove_consumer(provider_model)
             except Exception:
                 logger.exception("Failed to unregister provider as a consumer")
 
-        session.delete(provider)
+        try:
+            provider.clean_up()
+        except NotImplementedError:
+            logger.info(
+                "Being deleted provider of type %s does not have a clean_up method",
+                provider_model.type
+            )
+        except Exception as exc:
+            logger.exception(msg="Failed to clean up provider",
+                             extra={ "error_msg": str(exc) })
```

Comment:
Nice PR! actually `logger.exception` already logs the exception itself alone so no need for that extra param!

Reply:
Nice!
I'll fix,

Reply:
Fixed :)